### PR TITLE
Add internal wave test

### PR DESCRIPTION
**Pull request by cbegeman**
This PR ports the internal wave test group from legacy compass.

### CODE REVIEW

**Comment by cbegeman** [comment]
@xylar I broke something recently. I'll let you know when it's ready again.

**Comment by xylar** [comment]
No worries, I saw that this was in draft mode, so wasn't going to test it out until you gave the okay.

**Comment by cbegeman** [comment]
@xylar Actually, I think it's ready for review as-is. It failed earlier because I linked to wrong MPAS-O build. Feel free to give suggestions on the viz too. Right now it plots the initial and final states.

File: .gitignore
Comment (xylar) [inline .gitignore:13]: Just noting that we discussed this.  Ignoring PNG files causes trouble for the documentation so we don't want to include these in the `.gitignore.

File: compass/ocean/__init__.py
Comment (xylar) [inline compass/ocean/__init__.py:6]: I'm much less concerned about the imports but might as well keep them alphabetical, too.

File: docs/users_guide/ocean/test_groups/internal_wave.rst
Comment (xylar) [inline docs/users_guide/ocean/test_groups/internal_wave.rst:130]:
```suggestion
Since mixing is a strong function of horizontal viscosity, this test case performs 
20-day integrations of the model forward in time at 5 different values of the viscosity (with steps
```

File: docs/users_guide/ocean/test_groups/internal_wave.rst
Comment (xylar) [inline docs/users_guide/ocean/test_groups/internal_wave.rst:25]:
```suggestion
forward runs), which performs time integration of the model. There is an optional ``viz`` 
step which performs visualization of vertical cross-sections through the center of the domain.
```

File: docs/users_guide/ocean/test_groups/internal_wave.rst
Comment (xylar) [inline docs/users_guide/ocean/test_groups/internal_wave.rst:11]:
```suggestion
temperature has a linear background stratification with a sinusoidal perturbation in the center 
of the domain. This perturbation initiates symmetric waves that propage out from the center and 
then back to the center after meeting the solid boundaries. 
```

Comment (xylar) [reply docs/users_guide/ocean/test_groups/internal_wave.rst:11]: I'd like to keep the lines easier to read from the terminal, so under 80 characters or so.

File: compass/ocean/tests/internal_wave/namelist.forward
Comment (xylar) [inline compass/ocean/tests/internal_wave/namelist.forward:2]:
```suggestion
config_dt = '00:05:00'
config_btr_dt = '00:00:15'
```
How you have it works fine but it's a little unusual to have 3-digit hours.

File: compass/ocean/tests/internal_wave/ten_day_test/namelist.forward
Comment (xylar) [inline compass/ocean/tests/internal_wave/ten_day_test/namelist.forward:2]:
```suggestion
config_dt = '00:05:00'
config_btr_dt = '00:00:15'
```

File: compass/ocean/tests/internal_wave/ten_day_test/namelist.forward
Comment (xylar) [inline compass/ocean/tests/internal_wave/ten_day_test/namelist.forward:10]:
```suggestion
config_implicit_bottom_drag_coeff = 1.0e-2
config_use_cvmix_convection = .true.
config_AM_globalStats_enable = .true.
config_AM_globalStats_compute_on_startup = .true.
config_AM_globalStats_write_on_startup = .true.
config_AM_globalStats_text_file = .true.
```
You may have good reason for not including global stats in this run, but the legacy test case includes it and sometimes it's useful.  I'd include it if you don't have a reason not to.

File: compass/ocean/tests/internal_wave/ten_day_test/streams.forward
Comment (xylar) [inline compass/ocean/tests/internal_wave/ten_day_test/streams.forward:23]: Assuming you want global stats:
```suggestion

<stream name="globalStatsOutput"
        output_interval="0000_01:00:00"/>

</streams>
```

File: compass/ocean/tests/internal_wave/rpe_test/__init__.py
Comment (xylar) [inline compass/ocean/tests/internal_wave/rpe_test/__init__.py:32]: These are different from the viscosities in the legacy test case and in the original paper.  I think these might be copied from the baroclinic channel?

Following, https://github.com/MPAS-Dev/compass/blob/legacy/ocean/internal_waves/5km/rpe_test/plot.py#L10
```suggestion
        nus = [0.01, 1, 15, 150]
```
The formatting of the step names might get a little screwed up if `0.01` turns into something with 16 digits, so you might want to do some formatting magic there.

File: compass/ocean/tests/internal_wave/rpe_test/namelist.forward
Comment (xylar) [inline compass/ocean/tests/internal_wave/rpe_test/namelist.forward:1]: Just a note to say that I agree with reducing the run duration from 100 days to 20 days.  The analysis is performed at 20 days so no added advantage in running 80 more days by default.  Also, the experiments in the paper were run for 200 days, so 100 days isn't consistent with the paper either.

**Review by xylar — COMMENTED** [review]
Truly superb work as always.  All 3 test cases together ran perfectly on Ubuntu with gfortran in under 3 minutes.

My only major concern is that the viscosities for the RPE test case aren't the same as in the original test case or the paper.  Maybe that's just a question of having copied the viscosities from the baroclinic channel and not having updated them?

**Comment by xylar** [comment]
@mark-petersen, since you wrote the legacy test case, could you give this a look as soon as possible?  We'd ideally like to merge this one soon.

**Comment by xylar** [comment]
@cbegeman, when you get a chance, could you fix the PEP8 issues that the linter pointed out unless there's a good reason to keep the current formatting (e.g. long module names or URLs)?

File: compass/ocean/tests/internal_wave/rpe_test/__init__.py
Comment (xylar) [reply compass/ocean/tests/internal_wave/rpe_test/__init__.py:32]: I feel like I have a slight preference for sticking with the same set of nu values as in the paper, so the 4 values in my suggestion above.   That would make for an easier comparison with the model results from the paper if anyone wanted to do that. Maybe no one will actually want to do that and it's a moot point.

@mark-petersen, do you have an opinion?

**Comment by xylar** [comment]
@cbegeman, let me know when you'd like me to take another look.  It seems form your commits like you've addressed all my concerns.

**Comment by cbegeman** [comment]
@xylar You're welcome to take another look. I did make changes in response to all your suggestions. I figured I'd just wait for @mark-petersen's review.

**Comment by xylar** [comment]
Sounds good!

**Comment by mark-petersen** [comment]
Thank you @cbegeman for this helpful conversion from compass legacy. Tested with gnu optimized on grizzly. Created all plots after run. For example, here is a plot, required little fix I just pushed.
```
/lustre/scratch3/turquoise/mpeterse/runs/n/211004_internal_wave_1/ocean/internal_wave/rpe_test/analysis
sections_internal_wave.png
```
sections_internal_wave

File: compass/ocean/tests/internal_wave/namelist.forward
Comment (xylar) [inline compass/ocean/tests/internal_wave/namelist.forward:10]: @mark-petersen, you pushed this change as part of your commit.  I want to make sure it was intentional, since it wasn't mentioned in the commit message.

In general, I think it's a better practice to suggest changes as part of a review rather than pushing them yourself.

**Comment by cbegeman** [comment]
@mark-petersen Thanks for reviewing and catching the change needed to the viz script!

File: compass/ocean/tests/internal_wave/namelist.forward
Comment (mark-petersen) [reply compass/ocean/tests/internal_wave/namelist.forward]: You are right, that was unintentional. See comment below, I was testing that at the same time. I just corrected it, but I'll do the `suggest changes` method next time.

**Review by mark-petersen — APPROVED** [review]
Also run successfully with intel 19 on grizzly, fails with debug as noted above. I think the cvmix convection failure is out of scope for this PR, so I'm approving it. Thanks!

**Comment by mark-petersen** [comment]
I confirmed that my suggested fix in cvmix above,
```
components/mpas-ocean/src/cvmix/src/shared/cvmix_convection.F90:227
227                            CVmix_conv_params_in)
```
does fix this problem, i.e. I can run the test cases in this PR with gnu debug without errors.

**Comment by mark-petersen** [comment]
Issue reported: https://github.com/CVMix/CVMix-src/issues/92

**Review by xylar — APPROVED** [review]
I reran with the latest changes using gfortran on Ubuntu and everything looks good to me.

**Comment by xylar** [comment]
Thanks for the hard work on this one, @cbegeman!

**Pull request merged**
